npeps/npep-133-fqdn-egress-selector.md

```diff
@@ -0,0 +1,438 @@
+the existing Network Policy API, cannot be implemented currently. The design
+goal in this case is to ensure we do not make these unimplementable down the
+line.
+
+* As a cluster admin, I want to block all cluster egress traffic by default, and
+  require namespace admins to create NetworkPolicies explicitly allowing egress
+  to the domains they need to talk to.
+
+  The Cluster admin would use a `BaselineAdminNetworkPolicy` object to switch
+  the default disposition of the cluster. Namespace admins would then use a FQDN
+  selector in the Kubernetes `NetworkPolicy` objects to allow `my-service.com`.
+  
+## API
+
+This NPEP proposes adding a new type of `AdminNetworkPolicyEgressPeer` called
+`FQDNPeerSelector` which allows specifying domain names.
+
+```golang
+type AdminNetworkPolicyEgressPeer struct {
+    <snipped>
+    // FQDN provides a way to specify domain names as peers.
+    //
+    // FQDN is only supported for ALLOW rules. In order to control access, FQDN
+    // ALLOW rules should be used with a lower priority egress deny -- this
+    // allows the admin to maintain an explicit "allowlist" of reachable
+   // domains.
+    //
+    // Support: Extended
+    //
+    // +optional
+    FQDN *FQDNPeerSelector `json:"fqdn,omitempty"`
+}
+
+type FQDNPeerSelector struct {
+    // Name specifies the exact Fully-Qualified Domain Name to select (the
+    // trailing '.' is optional)
+    //
+    // Support: Extended
+    //
+    // +optional
+    // +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]([-a-zA-Z0-9_]*[a-zA-Z0-9])*\.?)*$`
+    Name string `json:"names,omitempty"`
+
+    //  Pattern allows matching FQDNs using wildcard specifiers.
+    //
+    // "*" matches 0 or more DNS valid characters (except for "."), and may
+    // occur anywhere in the pattern.
+    //  Examples:
+    //    - `*.kubernetes.io` matches subdomains of kubernetes.io at that level.
+    //      "www.kubernetes.io" and "blog.kubernetes.io" match, however
+    //      "kubernetes.io", "latest.blog.kubernetes.io", and "wikipedia.org" do
+    //      not.
+    //    - `*.*.kubernetes.io` matches subdomains of kubernetes.io at that
+    //      level.
+    //      "sub.subdomain.kubernetes.io" matches, however "kubernetes.io" and
+    //      "blog.kubernetes.io" do not.
+    //
+    // Support: Extended
+    //
+    // +optional
+    // +kubebuilder:validation:Pattern=`^([a-zA-Z0-9*]([-a-zA-Z0-9_*]*[a-zA-Z0-9*])*\.?)*$`
+    Pattern string `json:"pattern,omitempty"`
+}
+```
+
+### Examples
+
+#### Pods in `monitoring` namespace can talk to `my-service.com` and `*.cloud-provider.io`
+
+```yaml
+apiVersion: policy.networking.k8s.io/v1alpha1
+kind: AdminNetworkPolicy
+metadata:
+  name: allow-my-service-egress
+spec:
+  priority: 55
+  subject:
+    namespaces:
+      matchLabels:
+        kubernetes.io/metadata.name: "monitoring"
+  egress:
+  - name: "allow-to-my-service"
+    action: "Allow"
+    to:
+    - fqdn:
+        name: "my-service.com"
```

Comment:
is `host.subdomain.my-service.com` a match?

Reply:
It's not, the `name` field is supposed to be an exact match. I added some examples to clarify in the API docs.